### PR TITLE
WR-7820: fix unsetOffset method

### PR DESCRIPTION
**Pull request by ostrojen-wickedreports**

### CODE REVIEW

File: src/Collections/AbstractValidCollection.php [hunk 1]
```diff
@@ -2,84 +2,38 @@
...
-    }
-
-    public function key(): int
-    {
-        return $this->position;
-    }
-
-    public function valid(): bool
-    {
-        return $this->offsetExists($this->position);
```
Comment (ostrojen-wickedreports) [inline src/Collections/AbstractValidCollection.php:46]: There is a problem. When we call unsetOffest we delete collection key and valid method return false after deleted item

Comment (ostrojen-wickedreports) [reply src/Collections/AbstractValidCollection.php:46]: ArrayIterator has the correct logic for that, so I chose to use it.

File: src/Collections/AbstractValidCollection.php [hunk 2]
```diff
@@ -2,84 +2,38 @@
...
+     * field values validations here -> that would go against Single responsibility. Field values validation should be
+     * implemented in separate class/service.
+     */
+    abstract protected function isValidItem($item): bool;
 
-    public function count(): int
+    public function last()
     {
-        return count($this->collection);
+        $arrayCopy = $this->getArrayCopy();
```
Comment (moskeet-wickedreports) [inline src/Collections/AbstractValidCollection.php]: Please don't copy array.
This code will be used by our clients, and we should take into account memory issues as well!

Comment (moskeet-wickedreports) [reply src/Collections/AbstractValidCollection.php]:
```
function last(): void
{
    $keys = array_keys($$this->collection);
    $lastKey = last($keys);

    return $a[$lastKey];
}
```

**Review by moskeet-wickedreports — CHANGES_REQUESTED** [review]
Please try omit arrayCopy, due to memory issues

**Review by moskeet-wickedreports — APPROVED** [review]

**Pull request merged**